### ISSUE TITLE
signal 10 in 3.0.11 when used with sql and a custom condition

### ISSUE THREAD

**Issue by drook** [FreeRADIUS/freeradius-server#1832]
# Issue type
- [x] Defect - Crash or memory corruption.

# Defect/Feature description
signal 10 in 3.0.11 when used with sql and a custom condition, reproducible when a User is not a member of the Sql-Group. This custom condition is a workaround for #1833 . It's working for Sql-Users, but freeradius is crashing as soon as it encounters the user from users plain database.
## How to reproduce issue
Configure Sql database (in my case - postgresql), add some users and groups. Add the following condition to the post-auth section of the default site:

````post-auth condition
if(Sql-Group == "public-wifi"){
    update reply {
        Cisco-AVpair += "ssid=vivat"
    }
}
````
 
## Output of ``[radiusd|freeradius] -X`` showing issue occurring

```text
(19) Received Access-Request Id 134 from 127.0.0.1:30182 to 127.0.0.1:1812 length 83
(19)   User-Name = "pda"
(19)   User-Password = "XXXX"
(19)   NAS-IP-Address = 192.168.3.22
(19)   NAS-Port = 666
(19)   Message-Authenticator = 0x7026074d45652edc0ffdf6305d194053
(19)   EAP-Message = 0x0224000801706461
(19) # Executing section authorize from file /usr/local/etc/raddb/sites-enabled/default
(19)   authorize {
(19)     policy filter_username {
(19)       if (!&User-Name) {
(19)       if (!&User-Name)  -> FALSE
(19)       if (&User-Name =~ / /) {
(19)       if (&User-Name =~ / /)  -> FALSE
(19)       if (&User-Name =~ /@.*@/ ) {
(19)       if (&User-Name =~ /@.*@/ )  -> FALSE
(19)       if (&User-Name =~ /\\.\\./ ) {
(19)       if (&User-Name =~ /\\.\\./ )  -> FALSE
(19)       if ((&User-Name =~ /@/) && (&User-Name !~ /@(.+)\\.(.+)$/))  {
(19)       if ((&User-Name =~ /@/) && (&User-Name !~ /@(.+)\\.(.+)$/))   -> FALSE
(19)       if (&User-Name =~ /\\.$/)  {
(19)       if (&User-Name =~ /\\.$/)   -> FALSE
(19)       if (&User-Name =~ /@\\./)  {
(19)       if (&User-Name =~ /@\\./)   -> FALSE
(19)     } # policy filter_username = notfound
(19)     [preprocess] = ok
(19)     [chap] = noop
(19)     [mschap] = noop
(19)     [digest] = noop
(19) suffix: Checking for suffix after "@"
(19) suffix: No '@' in User-Name = "pda", looking up realm NULL
(19) suffix: No such realm "NULL"
(19)     [suffix] = noop
(19) eap: Peer sent EAP Response (code 2) ID 36 length 8
(19) eap: EAP-Identity reply, returning 'ok' so we can short-circuit the rest of authorize
(19)     [eap] = ok
(19)   } # authorize = ok
(19) Found Auth-Type = eap
(19) # Executing group from file /usr/local/etc/raddb/sites-enabled/default
(19)   authenticate {
(19) eap: Peer sent packet with method EAP Identity (1)
(19) eap: Calling submodule eap_md5 to process data
(19) eap_md5: Issuing MD5 Challenge
(19) eap: Sending EAP Request (code 1) ID 37 length 22
(19) eap: EAP session adding &reply:State = 0xd2a69d95d283996a
(19)     [eap] = handled
(19)   } # authenticate = handled
(19) Using Post-Auth-Type Challenge
(19) Post-Auth-Type sub-section not found.  Ignoring.
(19) # Executing group from file /usr/local/etc/raddb/sites-enabled/default
(19) Sent Access-Challenge Id 134 from 127.0.0.1:1812 to 127.0.0.1:30182 length 0
(19)   EAP-Message = 0x01250016041035c64445964d715fb408bcbc4645033c
(19)   Message-Authenticator = 0x00000000000000000000000000000000
(19)   State = 0xd2a69d95d283996acd1c515b7b036167
(19) Finished request
Waking up in 0.7 seconds.
(20) Received Access-Request Id 172 from 127.0.0.1:30182 to 127.0.0.1:1812 length 115
(20)   User-Name = "pda"
(20)   User-Password = "XXXX"
(20)   NAS-IP-Address = 192.168.3.22
(20)   NAS-Port = 666
(20)   Message-Authenticator = 0x8816ee5a84f76b71764be9f79917ff80
(20)   State = 0xd2a69d95d283996acd1c515b7b036167
(20)   EAP-Message = 0x0225001604106164c8f3eb9ef7926d985115fdeb3f21
(20) session-state: No cached attributes
(20) # Executing section authorize from file /usr/local/etc/raddb/sites-enabled/default
(20)   authorize {
(20)     policy filter_username {
(20)       if (!&User-Name) {
(20)       if (!&User-Name)  -> FALSE
(20)       if (&User-Name =~ / /) {
(20)       if (&User-Name =~ / /)  -> FALSE
(20)       if (&User-Name =~ /@.*@/ ) {
(20)       if (&User-Name =~ /@.*@/ )  -> FALSE
(20)       if (&User-Name =~ /\\.\\./ ) {
(20)       if (&User-Name =~ /\\.\\./ )  -> FALSE
(20)       if ((&User-Name =~ /@/) && (&User-Name !~ /@(.+)\\.(.+)$/))  {
(20)       if ((&User-Name =~ /@/) && (&User-Name !~ /@(.+)\\.(.+)$/))   -> FALSE
(20)       if (&User-Name =~ /\\.$/)  {
(20)       if (&User-Name =~ /\\.$/)   -> FALSE
(20)       if (&User-Name =~ /@\\./)  {
(20)       if (&User-Name =~ /@\\./)   -> FALSE
(20)     } # policy filter_username = notfound
(20)     [preprocess] = ok
(20)     [chap] = noop
(20)     [mschap] = noop
(20)     [digest] = noop
(20) suffix: Checking for suffix after "@"
(20) suffix: No '@' in User-Name = "pda", looking up realm NULL
(20) suffix: No such realm "NULL"
(20)     [suffix] = noop
(20) eap: Peer sent EAP Response (code 2) ID 37 length 22
(20) eap: No EAP Start, assuming it's an on-going EAP conversation
(20)     [eap] = updated
(20) files: users: Matched entry pda at line 57
(20)     [files] = ok
(20) sql: EXPAND %{User-Name}
(20) sql:    --> pda
(20) sql: SQL-User-Name set to 'pda'
rlm_sql (sql): Reserved connection (2)
(20) sql: EXPAND SELECT id, UserName, Attribute, Value, Op FROM radcheck WHERE Username = '%{SQL-User-Name}' ORDER BY id
(20) sql:    --> SELECT id, UserName, Attribute, Value, Op FROM radcheck WHERE Username = 'pda' ORDER BY id
(20) sql: Executing select query: SELECT id, UserName, Attribute, Value, Op FROM radcheck WHERE Username = 'pda' ORDER BY id
rlm_sql_postgresql: Status: PGRES_TUPLES_OK
rlm_sql_postgresql: query affected rows = 0 , fields = 5
(20) sql: EXPAND SELECT GroupName FROM radusergroup WHERE UserName='%{SQL-User-Name}' ORDER BY priority
(20) sql:    --> SELECT GroupName FROM radusergroup WHERE UserName='pda' ORDER BY priority
(20) sql: Executing select query: SELECT GroupName FROM radusergroup WHERE UserName='pda' ORDER BY priority
rlm_sql_postgresql: Status: PGRES_TUPLES_OK
rlm_sql_postgresql: query affected rows = 0 , fields = 1
(20) sql: User not found in any groups
rlm_sql (sql): Released connection (2)
(20)     [sql] = notfound
(20)     [expiration] = noop
(20)     [logintime] = noop
(20) noresetcounter: WARNING: Couldn't find check attribute, control:Max-All-Session, doing nothing...
(20)     [noresetcounter] = noop
(20) pap: WARNING: Auth-Type already set.  Not setting to PAP
(20)     [pap] = noop
(20)   } # authorize = updated
(20) Found Auth-Type = eap
(20) # Executing group from file /usr/local/etc/raddb/sites-enabled/default
(20)   authenticate {
(20) eap: Expiring EAP session with state 0xf84e4cd4f84456d8
(20) eap: Finished EAP session with state 0xd2a69d95d283996a
(20) eap: Previous EAP request found for state 0xd2a69d95d283996a, released from the list
(20) eap: Peer sent packet with method EAP MD5 (4)
(20) eap: Calling submodule eap_md5 to process data
(20) eap: Sending EAP Success (code 3) ID 37 length 4
(20) eap: Freeing handler
(20)     [eap] = ok
(20)   } # authenticate = ok
(20) # Executing section post-auth from file /usr/local/etc/raddb/sites-enabled/default
(20)   post-auth {
(20) reply_log: EXPAND /var/log/radacct/%{%{Packet-Src-IP-Address}:-%{Packet-Src-IPv6-Address}}/reply-detail-%Y%m%d
(20) reply_log:    --> /var/log/radacct/127.0.0.1/reply-detail-20161109
(20) reply_log: /var/log/radacct/%{%{Packet-Src-IP-Address}:-%{Packet-Src-IPv6-Address}}/reply-detail-%Y%m%d expands to /var/log/radacct/127.0.0.1/reply-detail-20161109
(20) reply_log: EXPAND %t
(20) reply_log:    --> Wed Nov  9 09:55:15 2016
(20)     [reply_log] = ok
(20) sql: EXPAND .query
(20) sql:    --> .query
(20) sql: Using query template 'query'
rlm_sql (sql): Reserved connection (3)
(20) sql: EXPAND %{User-Name}
(20) sql:    --> pda
(20) sql: SQL-User-Name set to 'pda'
(20) sql: EXPAND INSERT INTO radpostauth (username, pass, reply, authdate) VALUES('%{User-Name}', '%{%{User-Password}:-Chap-Password}', '%{reply:Packet-Type}', NOW())
(20) sql:    --> INSERT INTO radpostauth (username, pass, reply, authdate) VALUES('pda', 'XXXX', 'Access-Accept', NOW())
(20) sql: Executing query: INSERT INTO radpostauth (username, pass, reply, authdate) VALUES('pda', 'XXXX', 'Access-Accept', NOW())
rlm_sql_postgresql: Status: PGRES_COMMAND_OK
rlm_sql_postgresql: query affected rows = 1
(20) sql: SQL query returned: success
(20) sql: 1 record(s) updated
rlm_sql (sql): Released connection (3)
(20)     [sql] = ok
(20)     [exec] = noop
(20)     if (Sql-Group == "public-wifi"){
(20)     sql_groupcmp
(20)     EXPAND %{User-Name}
(20)        --> pda
(20)     SQL-User-Name set to 'pda'
rlm_sql (sql): Reserved connection (4)
(20)     EXPAND SELECT GroupName FROM radusergroup WHERE UserName='%{SQL-User-Name}' ORDER BY priority
(20)        --> SELECT GroupName FROM radusergroup WHERE UserName='pda' ORDER BY priority
(20)     Executing select query: SELECT GroupName FROM radusergroup WHERE UserName='pda' ORDER BY priority
rlm_sql_postgresql: Status: PGRES_TUPLES_OK
rlm_sql_postgresql: query affected rows = 0 , fields = 1
rlm_sql (sql): Released connection (4)
(20)     sql_groupcmp finished: User is NOT a member of group public-wifi
zsh: bus error  radiusd -X
```

dmesg (I have a watchdog written for freeradiusd from the times when it was unstable at 2.x):

````dmesg
# dmesg -a
pid 71793 (radiusd), uid 133: exited on signal 10
pid 71847 (radiusd), uid 133: exited on signal 10
pid 71874 (radiusd), uid 133: exited on signal 10
pid 71931 (radiusd), uid 133: exited on signal 10
pid 71992 (radiusd), uid 133: exited on signal 10
pid 72049 (radiusd), uid 133: exited on signal 10
pid 72112 (radiusd), uid 133: exited on signal 10
pid 72143 (radiusd), uid 133: exited on signal 10
pid 72206 (radiusd), uid 133: exited on signal 10
pid 72220 (radiusd), uid 133: exited on signal 10
pid 72282 (radiusd), uid 133: exited on signal 10
pid 72360 (radiusd), uid 133: exited on signal 10
pid 72427 (radiusd), uid 133: exited on signal 10
pid 72514 (radiusd), uid 133: exited on signal 10
pid 72591 (radiusd), uid 133: exited on signal 10
pid 72651 (radiusd), uid 133: exited on signal 10
pid 73659 (radiusd), uid 133: exited on signal 10
pid 73725 (radiusd), uid 133: exited on signal 10
pid 75192 (radiusd), uid 133: exited on signal 10
pid 75322 (radiusd), uid 133: exited on signal 10
pid 75400 (radiusd), uid 133: exited on signal 10
pid 75514 (radiusd), uid 133: exited on signal 10
pid 75574 (radiusd), uid 133: exited on signal 10
pid 75653 (radiusd), uid 133: exited on signal 10
pid 75718 (radiusd), uid 133: exited on signal 10
````

**Comment by drook** [comment]
Just to prove that it's not my configs/unlang statements are broken, I repeated this on a test server installed and configured from scratch, got same results.

**Comment by alanbuxey** [comment]
(20) sql: SQL-User-Name set to 'pda'
rlm_sql (sql): Reserved connection (2)
(20) sql: EXPAND SELECT id, UserName, Attribute, Value, Op FROM
radcheck WHERE Username = '%{SQL-User-Name}' ORDER BY id
(20) sql:    --> SELECT id, UserName, Attribute, Value, Op FROM
radcheck WHERE Username = 'pda' ORDER BY id
(20) sql: Executing select query: SELECT id, UserName, Attribute,
Value, Op FROM radcheck WHERE Username = 'pda' ORDER BY id
rlm_sql_postgresql: Status: PGRES_TUPLES_OK
rlm_sql_postgresql: query affected rows = 0 , fields = 5
(20) sql: EXPAND SELECT GroupName FROM radusergroup WHERE
UserName='%{SQL-User-Name}' ORDER BY priority
(20) sql:    --> SELECT GroupName FROM radusergroup WHERE
UserName='pda' ORDER BY priority
(20) sql: Executing select query: SELECT GroupName FROM radusergroup
WHERE UserName='pda' ORDER BY priority
rlm_sql_postgresql: Status: PGRES_TUPLES_OK
rlm_sql_postgresql: query affected rows = 0 , fields = 1
(20) sql: User not found in any groups
rlm_sql (sql): Released connection (2)
(20)     [sql] = notfound

(20)     Executing select query: SELECT GroupName FROM radusergroup
WHERE UserName='pda' ORDER BY priority
rlm_sql_postgresql: Status: PGRES_TUPLES_OK
rlm_sql_postgresql: query affected rows = 0 , fields = 1



is this use in your SQL tables.....does this user have an entry in
your radusergroup table?  from the debug, there are no results being
found.

hence there is nothign populated.  either fix the SQL tables/entries
or adjust your group query in the postgresl sql.conf ?


alan

**Comment by arr2036** [comment]
This was fixed in 3.0.12

**Comment by drook** [comment]
@alanbuxey , this user is from users text database, it's not from SQL. As far as I understand I don't have to store all of my users in just one backend, I'm free to use whatever backend I want.
